### ISSUE TITLE
KPI 32 calculation query

### ISSUE THREAD

**Issue by anchit-chandran** [rcpch/national-paediatric-diabetes-audit#274]
What counts as eligible for KPI 32?

![Image](https://github.com/user-attachments/assets/1eb40f93-b7a0-448d-8eb5-bd00d6c5f8c3)

The eligibility for KPIS 25-29,31 include:
- eligible for KPI 5 patients (KPIS 25-27)
- eligible for KPI 6 patients (KPIS 28,29,31)

So I thought it would just be the union of the criteria for KPI5 AND 6:

**CRITERIA FOR BOTH:**
```python
Total number of patients with:
* a valid NHS number
* Diagnosis of Type 1 diabetes

Excluding
* Date of diagnosis within the audit period
* Date of leaving service within the audit period
* Date of death within the audit period
```

**KPI 5:**
```python
Total number of patients with:
*a valid date of birth
*a valid PDU number
* a visit date or admission date within the audit period
* Below the age of 25 at the start of the audit period
``` 

**KPI 6:**
```python
Total number of patients with:
* an observation within the audit period
* Age 12 and above at the start of the audit period
```

A child with who's 10 years old qualifies for KPI5 but not for KPI6. So union'ing might not be correct here? Unsure how to calculate this

**Comment by eatyourpeas** [comment]
What this is saying I think that KPI32 is measuring all eligible children (ie T1DM, not T2DM/CFRD/MODY etc) who have had a complete annual review. The complication is that the contents of the annual review depend on your age: you do not need to do a foot check, an ACR, a blood pressure or retinal DECS screening in the under 12s. Why they pick 12y as some magic number is beyond me - it is not like your feet somehow change at 12y (though puberty is in the post by then so you will go up a few shoe sizes). Anyway, this means you would have to run four queries I think: two to get two different denominators (eligible patients >=12y, eligible patients < 12y), and two to get two different nominators (those for the >12s and those for the under 12s). You then sum the totals of each together.
Hopefully @cillian-rcpch will have a query he can show us

**Comment by AmaniKrayemRCPCH** [comment]
The way we currently do it is to calculate health check completion rate for those <12 years (HbA1c, BMI, and thyroid screen), and health check completion rate for those >12 years (HbA1c, BMI, thyroid screen, blood pressure, urinary albumin, and foot exam). 

The total health check completion rate is those two added together. 

Does that make sense? @cillian-rcpch do you agree?

**Comment by cillian-rcpch** [comment]
Hi, sorry that I missed this thread until now. It's pretty much as Simon and Amani have said.

Health check completion rate is calculated in the following way:
![image](https://github.com/user-attachments/assets/223846a1-665b-48cb-ba52-4d6406c0323b)

It only considers children and young people with Type 1 diabetes who have had a "complete year of care". 
That means the CYP was not diagnosed, did not transition or did not die within the audit period. 

In plain language the rate just represents the percentage of health checks which should be given, that actually ARE given (it should be 100% in a perfect world). 

Hopefully that answers it but let me know if still unclear.

**Comment by anchit-chandran** [comment]
thanks all, this makes sense - will confirm in sprint later and update excel sheet definition so clearer + be centralised/version tracked

**Comment by anchit-chandran** [comment]
@AmaniKrayemRCPCH to update on definitions

**Comment by AmaniKrayemRCPCH** [comment]
So KPI 32 looks at all the health checks completed, out of all the possible health checks. So, if a unit has 10 kids <12, and 20 kids >=12, we would expect them to have completed 150 health checks each year (10 x 3, 20 x 6). 

KPI 32a and 32b look at the number of individual children who've received all their health checks. 

```
KPI 32 - Health Check Completion Rate (Percentage of total expected health checks completed)
numerator = number of health checks given
denominator = number of expected health checks (3 for CYP <12 years with T1D, 6 for CYP >= 12 years with T1S)

KPI 32a - Health Checks (Less than 12 years) 
numerator = number of CYP with T1D under 12 years with all three health checks (HbA1c, BMI, Thyroid)
denominator  = number of CYP with T1D under 12 years

KPI 32b - Health Checks (12 years and over)
numerator = Number of CYP with T1D aged 12 years and over with all six health checks (HbA1c, BMI, Thyroid, BP, Urinary Albumin, Foot Exam)
denominator = Number of CYP with T1D aged 12 years and over```

**Comment by AmaniKrayemRCPCH** [comment]
I can update the excel sheet if that's clearer

**Comment by anchit-chandran** [comment]
have updated on excel sheet like so, can update if needed, will be in main branch once I merge in kpi stuff.

internally, calling them KPI 32.1-3; if they need to be labelled `32`, `32A`, `32B`, we can do that on the front end

![Image](https://github.com/user-attachments/assets/51ead433-9399-4196-9804-bf9bcaa84f48)

**Comment by anchit-chandran** [comment]
@eatyourpeas more of a data model q - how do we find patients with a "complete year of care"? 

The definition for KPI5 doesn't have anything explicit checking complete year of care so this is how I've been calculating it in `kpis.py`:
```python


    def calculate_kpi_5_total_t1dm_complete_year(self) -> KPIResult:
        """
        Calculates KPI 5: Total number of patients with T1DM who have completed a year of care
        "Total number of patients with:
        * a valid NHS number
        *a valid date of birth
        *a valid PDU number
        * a visit date or admission date within the audit period
        * Below the age of 25 at the start of the audit period* Diagnosis of Type 1 diabetes

        Excluding
        * Date of diagnosis within the audit period
        * Date of leaving service within the audit period
        * Date of death within the audit period"
        """
        # If we have not already calculated KPI 1, do so now to set
        total_kpi_1_eligible_pts_base_query_set, total_eligible_kpi_1 = (
            self._get_total_kpi_1_eligible_pts_base_query_set_and_total_count()
        )

        eligible_patients = total_kpi_1_eligible_pts_base_query_set.exclude(
            # EXCLUDE Date of diagnosis within the audit period
            Q(diagnosis_date__range=(self.AUDIT_DATE_RANGE))
            # EXCLUDE Date of leaving service within the audit period
            | (
                Q(
                    paediatric_diabetes_units__date_leaving_service__range=(
                        self.audit_start_date,
                        self.audit_end_date,
                    )
                )
            )
            # EXCLUDE Date of death within the audit period"
            | Q(death_date__range=(self.AUDIT_DATE_RANGE))
        ).distinct()
```

**Comment by anchit-chandran** [comment]
bump on this! @eatyourpeas

**Comment by eatyourpeas** [comment]
Oops I am sorry.
Yes that is to my mind the way to calculate it. If date of diagnosis does not fall within the audit year and all the other eligibility items apply then that is a complete year of care

**Comment by anchit-chandran** [comment]
FYI assumptions for calculation @cillian-rcpch  @AmaniKrayemRCPCH 

- 'BMI' health check inferred from `height` & `weight` & `height_weight_observation_date` all being present (@eatyourpeas we don't store anything directly for BMI right?)
- A health check is counted as passed by looking across ALL `Visits` for a given patient and given audit date range -> if ANY of these have a value for the health check, then it's counted. E.g. a `thyroid_function` health check is counted if Patient A has the following visit values:

```
Visit 1 -> thyroid_function_date = None
Visit 2 -> thyroid_function_date = None
Visit 3 -> thyroid_function_date = 1 Jan 2024
Visit 4 -> thyroid_function_date = None
```

If all ok, can close as implemented in #301

**Comment by eatyourpeas** [comment]
Yes - BMI is a calculated field from height and weight. It strikes me I need to raise a new issue here as we should use the RCPCHGrowth API to calculate SDS for BMI and signpost this in the UI

Yes - a health check counts as one of the measures performed across any visit in the audit period. I will close this and post a new issue for the BMI
